notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/NotificationHubInstallationAdapter.java

```diff
@@ -0,0 +1,148 @@
+     */
+    @Override
+    public void saveInstallation(Context context, Installation installation) {
+        RequestQueue queue = Volley.newRequestQueue(context.getApplicationContext());
+
+        String formatEndpoint = NotificationHubInstallationHelper.parseSbEndpoint(mConnectionString.getEndpoint());
+        String url = NotificationHubInstallationHelper.getInstallationUrl(formatEndpoint, mHubName, installation.getInstallationId());
+
+        StringRequest request = new StringRequest(
+                Request.Method.PUT,
+                url,
+                response -> {
+                    Log.i("ANH", "Installation Updated");
+                },
+                error -> {
+                    Log.e("ANH", "Couldn't update installation: " + error);
+                }
+        ){
+            @Override
+            public byte[] getBody() {
+                try {
+                    JSONArray tagList = new JSONArray();
+                    for (String tag: installation.getTags()) {
+                        tagList.put(tag);
+                    }
+
+                    JSONObject jsonBody = new JSONObject(){{
+                        put("installationId", installation.getInstallationId());
+                        put("platform", "GCM");
+                        put("pushChannel", installation.getPushChannel());
+                        put("tags", tagList);
+                    }};
+                    return jsonBody.toString().getBytes(StandardCharsets.UTF_8);
+                } catch (JSONException e) {
+                    e.printStackTrace();
+                    return null;
+                }
+            }
+
+            @Override
+            public  Map<String, String> getHeaders(){
+                try {
+                    Map<String,String> params = new HashMap<String, String>(){{
+                        put("Content-Type", "application/json");
+                        put("x-ms-version", "2015-01");
+                        put("Authorization", generateAuthToken(url));
+                    }};
+                    return params;
+                } catch (InvalidKeyException e) {
+                    e.printStackTrace();
+                    return null;
+                }
+            }
+        };
+
+        queue.add(request);
+    }
+
+
+    private String generateAuthToken(String url) throws InvalidKeyException {
+        String keyName = mConnectionString.getSharedAccessKeyName();
+        String key = mConnectionString.getSharedAccessKey();
+
+        try {
+            url = URLEncoder.encode(url, "UTF-8").toLowerCase(Locale.ENGLISH);
+        } catch (UnsupportedEncodingException e) {
+            // this shouldn't happen because of the fixed encoding
+        }
+
+        // Set expiration in seconds
+        Calendar expireDate = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+        expireDate.add(Calendar.MINUTE, 5);
+        long expires = expireDate.getTimeInMillis() / 1000;

```

Comment:
I think this can be simplified to:
```
long expires = (System.currentTimeMillis() / 1000) + EXPIRE_SECONDS;
```

`System.currentTimeMillis()` returns an epoch time (which is already in UTC).

Reply:
Fixed in 0fd9db171f16512a14fb8c49f13afa6eb1782ffd